File: Apps/Android/app/src/main/kotlin/com/opencapture/openpocketcine/ChromeEditChrome.kt
package com.opencapture.openpocketcine

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import kotlin.math.max
import kotlin.math.min

private const val BADGE = 26f
private const val BADGE_GAP = 4f
private const val EDGE_INSET = 16f
private const val OVERHANG = 4f

@Composable
fun ChromeEditBanner(
    mode: PocketDispMode,
    modifier: Modifier = Modifier,
    onDone: () -> Unit,
) {
    Row(
        modifier
            .overlayGlass(RoundedCornerShape(50))
            .padding(start = 12.dp, end = 6.dp, top = 6.dp, bottom = 6.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Canvas(Modifier.size(12.dp)) {
            drawEyeGlyph(LiveDesign.accent, slashed = false)
        }
        Column(verticalArrangement = Arrangement.spacedBy(1.dp)) {
            Text("Editing ${mode.title}", color = LiveDesign.text, style = LiveType.ui(11.5f, FontWeight.SemiBold))
            Text("Tap an eye to show or hide it", color = LiveDesign.muted, style = LiveType.ui(10f))
        }
        Box(
            Modifier
                .clip(RoundedCornerShape(50))
                .background(LiveDesign.accent)
                .chromeClickable(onClick = onDone)
                .padding(horizontal = 12.dp, vertical = 6.dp),
        ) {
            Text("Done", color = LiveDesign.background, style = LiveType.ui(11.5f, FontWeight.Bold))
        }
    }
}

@Composable
fun ChromeEditBadgeLayer(
    mode: PocketDispMode,
    boxes: List<Pair<PocketDispSection, ChromeRect>>,
    viewportWidth: Float,
    viewportHeight: Float,
    visible: (PocketDispSection) -> Boolean,
    onToggle: (PocketDispSection) -> Unit,
) {
    val frames = badgeFrames(boxes, viewportWidth, viewportHeight)
    frames.forEach { (section, frame) ->
        val on = visible(section)
        Box(
            Modifier
                .offset(frame.x.dp, frame.y.dp)
                .size(BADGE.dp)
                .clip(CircleShape)
                .background(if (on) LiveDesign.accent else Color.Black.copy(alpha = 0.9f))
                .border(1.dp, LiveDesign.text.copy(alpha = 0.55f), CircleShape)
                .chromeClickable(onClick = { onToggle(section) }),
            contentAlignment = Alignment.Center,
        ) {
            Canvas(Modifier.size(12.dp)) {
                drawEyeGlyph(
                    color = if (on) LiveDesign.background else LiveDesign.text,
                    slashed = !on,
                    filled = true,
                    pupil = if (on) LiveDesign.accent else Color.Black,
                )
            }
        }
    }
}

fun Modifier.chromeEditStroke(editing: Boolean, visible: Boolean): Modifier {
    if (!editing) return this
    val color = if (visible) LiveDesign.accent.copy(alpha = 0.75f) else LiveDesign.muted.copy(alpha = 0.75f)
    return drawWithContent {
        drawContent()
        drawRoundRect(
            color = color,
            cornerRadius = CornerRadius(8.dp.toPx()),
            style = Stroke(width = 1.dp.toPx(), pathEffect = PathEffect.dashPathEffect(floatArrayOf(6f, 6f))),
        )
    }
}

internal fun chromeEditBoxes(
    layout: LiveMonitorLayout,
    model: AppModel,
    uiLocked: Boolean,
    zoom: ChromeRect,
    stick: ChromeRect,
    statusChips: Map<PocketDispSection, ChromeRect> = emptyMap(),
): List<Pair<PocketDispSection, ChromeRect>> {
    val out = mutableListOf<Pair<PocketDispSection, ChromeRect>>()
    fun add(section: PocketDispSection, rect: ChromeRect) {
        if (!rect.isEmpty) out += section to rect
    }
    if (model.chromeSectionMounts(PocketDispSection.STATUS_BAR)) add(PocketDispSection.STATUS_BAR, layout.topDeck)
    for (section in TOP_STATUS_CHIPS) {
        if (model.chromeSectionMounts(section)) statusChips[section]?.let { add(section, it) }
    }
    if (model.chromeSectionMounts(PocketDispSection.LOCK_BUTTON) || uiLocked) add(PocketDispSection.LOCK_BUTTON, layout.lock)
    if (model.chromeSectionMounts(PocketDispSection.BATTERIES)) add(PocketDispSection.BATTERIES, layout.battery)
    if (model.chromeSectionMounts(PocketDispSection.RAIL_SETTINGS)) add(PocketDispSection.RAIL_SETTINGS, layout.settings)
    if (model.chromeSectionMounts(PocketDispSection.RAIL_MEDIA)) add(PocketDispSection.RAIL_MEDIA, layout.media)
    if (model.chromeSectionMounts(PocketDispSection.RAIL_RECORD)) add(PocketDispSection.RAIL_RECORD, layout.record)
    if (model.chromeSectionMounts(PocketDispSection.TOOL_BAR)) add(PocketDispSection.TOOL_BAR, layout.assist)
    if (model.chromeSectionMounts(PocketDispSection.CAMERA_VALUES)) add(PocketDispSection.CAMERA_VALUES, layout.capture)
    if (model.chromeSectionMounts(PocketDispSection.ZOOM_CHIP)) add(PocketDispSection.ZOOM_CHIP, zoom)
    if (model.chromeSectionMounts(PocketDispSection.GIMBAL_STICK)) add(PocketDispSection.GIMBAL_STICK, stick)
    return out
}

internal val TOP_STATUS_CHIPS =
    listOf(
        PocketDispSection.REC_READOUT,
        PocketDispSection.TIMECODE,
        PocketDispSection.FORMAT,
        PocketDispSection.COLOR,
        PocketDispSection.STORAGE,
        PocketDispSection.FPS,
    )

private enum class BadgeAnchor {
    TOP_LEADING,
    TOP,
    TOP_TRAILING,
    LEADING,
    TRAILING,
    BOTTOM_LEADING,
    BOTTOM,
    BOTTOM_TRAILING,
    ;

    fun frame(box: ChromeRect): ChromeRect {
        val inner = BADGE - OVERHANG
        val midX = box.midX - BADGE / 2f
        val midY = box.midY - BADGE / 2f
        val x =
            when (this) {
                TOP_LEADING, LEADING, BOTTOM_LEADING -> box.minX - OVERHANG
                TOP, BOTTOM -> midX
                TOP_TRAILING, TRAILING, BOTTOM_TRAILING -> box.maxX - inner
            }
        val y =
            when (this) {
                TOP_LEADING, TOP, TOP_TRAILING -> box.minY - OVERHANG
                LEADING, TRAILING -> midY
                BOTTOM_LEADING, BOTTOM, BOTTOM_TRAILING -> box.maxY - inner
            }
        return ChromeRect(x, y, BADGE, BADGE)
    }
}

private fun badgeFrames(
    boxes: List<Pair<PocketDispSection, ChromeRect>>,
    vw: Float,
    vh: Float,
): List<Pair<PocketDispSection, ChromeRect>> {
    val playable =
        ChromeRect(
            min(EDGE_INSET, max(0f, vw - BADGE)),
            min(EDGE_INSET, max(0f, vh - BADGE)),
            max(BADGE, vw - 2f * EDGE_INSET),
            max(BADGE, vh - 2f * EDGE_INSET),
        )
    val placed = mutableListOf<ChromeRect>()
    val result = mutableListOf<Pair<PocketDispSection, ChromeRect>>()
    for ((section, box) in boxes) {
        if (box.isEmpty) continue
        val preferTrailing = box.midX < vw / 2f
        val preferBottom = box.midY < vh / 2f
        val candidates =
            BadgeAnchor.entries
                .sortedBy { rank(it, preferTrailing, preferBottom) }
                .map { clampBadge(it.frame(box), playable) }
        val choice =
            candidates.firstOrNull { candidate ->
                placed.none { overlaps(it, candidate) }
            } ?: candidates.first()
        placed += choice
        result += section to choice
    }
    return result
}

private fun clampBadge(frame: ChromeRect, playable: ChromeRect): ChromeRect {
    val x = min(max(frame.minX, playable.minX), playable.maxX - frame.width)
    val y = min(max(frame.minY, playable.minY), playable.maxY - frame.height)
    return ChromeRect(x, y, frame.width, frame.height)
}

private fun overlaps(a: ChromeRect, b: ChromeRect): Boolean =
    a.minX < b.maxX + BADGE_GAP &&
        b.minX < a.maxX + BADGE_GAP &&
        a.minY < b.maxY + BADGE_GAP &&
        b.minY < a.maxY + BADGE_GAP

private fun rank(anchor: BadgeAnchor, preferTrailing: Boolean, preferBottom: Boolean): Int {
    val horizontal =
        when (anchor) {
            BadgeAnchor.TOP_TRAILING, BadgeAnchor.TRAILING, BadgeAnchor.BOTTOM_TRAILING ->
                if (preferTrailing) 0 else 2
            BadgeAnchor.TOP_LEADING, BadgeAnchor.LEADING, BadgeAnchor.BOTTOM_LEADING ->
                if (preferTrailing) 2 else 0
            BadgeAnchor.TOP, BadgeAnchor.BOTTOM -> 1
        }
    val vertical =
        when (anchor) {
            BadgeAnchor.BOTTOM_LEADING, BadgeAnchor.BOTTOM, BadgeAnchor.BOTTOM_TRAILING ->
                if (preferBottom) 0 else 2
            BadgeAnchor.TOP_LEADING, BadgeAnchor.TOP, BadgeAnchor.TOP_TRAILING ->
                if (preferBottom) 2 else 0
            BadgeAnchor.LEADING, BadgeAnchor.TRAILING -> 1
        }
    return horizontal + vertical
}

private fun DrawScope.drawEyeGlyph(
    color: Color,
    slashed: Boolean,
    filled: Boolean = false,
    pupil: Color = color,
) {
    val cx = size.width / 2f
    val cy = size.height / 2f
    val w = size.width * 0.78f
    val h = size.height * 0.42f
    val outline =
        Path().apply {
            moveTo(cx - w / 2f, cy)
            quadraticTo(cx, cy - h, cx + w / 2f, cy)
            quadraticTo(cx, cy + h, cx - w / 2f, cy)
            close()
        }
    val stroke =
        Stroke(
            width = 1.35.dp.toPx(),
            cap = StrokeCap.Round,
            join = StrokeJoin.Round,
        )
    if (filled) {
        drawPath(outline, color)
        drawCircle(pupil, radius = size.minDimension * 0.13f, center = Offset(cx, cy))
    } else {
        drawPath(outline, color, style = stroke)
        drawCircle(color, radius = size.minDimension * 0.12f, center = Offset(cx, cy))
    }
    if (slashed) {
        drawLine(
            color,
            Offset(size.width * 0.18f, size.height * 0.82f),
            Offset(size.width * 0.82f, size.height * 0.18f),
            strokeWidth = 1.5.dp.toPx(),
            cap = StrokeCap.Round,
        )
    }
}

File: ios/OpenPocketCine/ChromeEditChrome.swift
import SwiftUI

// MARK: - Edit view (per-DISP chrome)

/// Measured bounds of every badgeable monitor element, in canvas space.
///
/// Boxes are measured, never derived from the zone map: a zone frame is a layout
/// budget, and several elements draw inset within theirs.
struct ChromeEditBoundsKey: PreferenceKey {
    static let defaultValue: [PocketDispChrome.Section: CGRect] = [:]

    static func reduce(
        value: inout [PocketDispChrome.Section: CGRect],
        nextValue: () -> [PocketDispChrome.Section: CGRect]
    ) {
        value.merge(nextValue()) { existing, next in existing.union(next) }
    }
}

enum ChromeEditSpace {
    static var name: String { LiveCanvasSpace.name }
}

/// Where the Edit view puts each element's eye badge. Pure geometry so tests can
/// pin placement without a live monitor.
///
/// Badges sit **on** the control, on the side toward the canvas centre. Centering
/// them on an outer corner (then clamping 3 pt in) put rail / top-deck eyes in
/// the display corner radius, where they were clipped and untappable.
enum PocketChromeEditLayout {
    static let badgeSize: CGFloat = 26
    static let badgeGap: CGFloat = 4
    /// Hard inset from the physical canvas. 16 pt keeps a 26 pt eye inside the
    /// iPhone display corner curve (~52 pt).
    static let edgeInset: CGFloat = 16
    /// How far a badge may peek past its element's edge. The rest sits on it.
    static let overhang: CGFloat = 4

    struct Box: Equatable, Sendable {
        let section: PocketDispChrome.Section
        let frame: CGRect
    }

    static func badgeFrames(
        _ boxes: [Box],
        viewport: CGSize,
        badgeSize: CGFloat = badgeSize
    ) -> [PocketDispChrome.Section: CGRect] {
        let playable = playableRect(in: viewport, badgeSize: badgeSize)
        var placed: [CGRect] = []
        var result: [PocketDispChrome.Section: CGRect] = [:]

        for box in boxes where box.frame.width > 1 && box.frame.height > 1 {
            let candidates = anchors(on: box.frame, viewport: viewport, badgeSize: badgeSize)
                .map { clamped($0, in: playable) }
            let choice =
                candidates.first { candidate in
                    !placed.contains { overlaps($0, candidate) }
                } ?? candidates[0]
            placed.append(choice)
            result[box.section] = choice
        }
        return result
    }

    static func playableRect(in viewport: CGSize, badgeSize: CGFloat = badgeSize) -> CGRect {
        let inset = edgeInset
        let width = max(badgeSize, viewport.width - 2 * inset)
        let height = max(badgeSize, viewport.height - 2 * inset)
        let x = min(inset, max(0, viewport.width - width))
        let y = min(inset, max(0, viewport.height - height))
        return CGRect(x: x, y: y, width: width, height: height)
    }

    /// 4 corners + 4 mid-edges, badge mostly overlapping `frame`, inward first.
    private static func anchors(
        on frame: CGRect,
        viewport: CGSize,
        badgeSize: CGFloat
    ) -> [CGRect] {
        let preferTrailing = frame.midX < viewport.width / 2
        let preferBottom = frame.midY < viewport.height / 2
        return Anchor.allCases
            .sorted {
                rank($0, preferTrailing: preferTrailing, preferBottom: preferBottom)
                    < rank($1, preferTrailing: preferTrailing, preferBottom: preferBottom)
            }
            .map { $0.frame(on: frame, badgeSize: badgeSize, overhang: overhang) }
    }

    private enum Anchor: CaseIterable {
        case topLeading, top, topTrailing
        case leading, trailing
        case bottomLeading, bottom, bottomTrailing

        func frame(on box: CGRect, badgeSize: CGFloat, overhang: CGFloat) -> CGRect {
            let inner = badgeSize - overhang
            let midX = box.midX - badgeSize / 2
            let midY = box.midY - badgeSize / 2
            let x: CGFloat
            let y: CGFloat
            switch self {
            case .topLeading:
                x = box.minX - overhang
                y = box.minY - overhang
            case .top:
                x = midX
                y = box.minY - overhang
            case .topTrailing:
                x = box.maxX - inner
                y = box.minY - overhang
            case .leading:
                x = box.minX - overhang
                y = midY
            case .trailing:
                x = box.maxX - inner
                y = midY
            case .bottomLeading:
                x = box.minX - overhang
                y = box.maxY - inner
            case .bottom:
                x = midX
                y = box.maxY - inner
            case .bottomTrailing:
                x = box.maxX - inner
                y = box.maxY - inner
            }
            return CGRect(x: x, y: y, width: badgeSize, height: badgeSize)
        }
    }

    /// 0 = most inward (toward canvas centre), 4 = most outward (screen corner).
    private static func rank(
        _ anchor: Anchor,
        preferTrailing: Bool,
        preferBottom: Bool
    ) -> Int {
        let horizontal: Int
        switch anchor {
        case .topTrailing, .trailing, .bottomTrailing:
            horizontal = preferTrailing ? 0 : 2
        case .topLeading, .leading, .bottomLeading:
            horizontal = preferTrailing ? 2 : 0
        case .top, .bottom:
            horizontal = 1
        }
        let vertical: Int
        switch anchor {
        case .bottomLeading, .bottom, .bottomTrailing:
            vertical = preferBottom ? 0 : 2
        case .topLeading, .top, .topTrailing:
            vertical = preferBottom ? 2 : 0
        case .leading, .trailing:
            vertical = 1
        }
        return horizontal + vertical
    }

    private static func overlaps(_ a: CGRect, _ b: CGRect) -> Bool {
        a.minX < b.maxX + badgeGap && b.minX < a.maxX + badgeGap
            && a.minY < b.maxY + badgeGap && b.minY < a.maxY + badgeGap
    }

    private static func clamped(_ frame: CGRect, in playable: CGRect) -> CGRect {
        guard playable.width >= frame.width, playable.height >= frame.height else {
            return CGRect(
                x: playable.minX, y: playable.minY, width: frame.width, height: frame.height)
        }
        return CGRect(
            x: min(max(frame.minX, playable.minX), playable.maxX - frame.width),
            y: min(max(frame.minY, playable.minY), playable.maxY - frame.height),
            width: frame.width,
            height: frame.height
        )
    }
}

extension View {
    func chromeEditable(
        _ section: PocketDispChrome.Section,
        editing mode: PocketDispMode?
    ) -> some View {
        modifier(ChromeEditable(section: section, mode: mode))
    }
}

struct ChromeEditable: ViewModifier {
    @Environment(AppModel.self) private var model
    let section: PocketDispChrome.Section
    let mode: PocketDispMode?

    func body(content: Content) -> some View {
        if let mode, PocketDispChrome.isConfigurable(section, in: mode) {
            let on = model.chrome(for: mode).isVisible(section)
            content
                .opacity(on ? 1 : 0.3)
                .background {
                    GeometryReader { proxy in
                        Color.clear.preference(
                            key: ChromeEditBoundsKey.self,
                            value: [section: proxy.frame(in: .named(ChromeEditSpace.name))]
                        )
                    }
                }
                .overlay {
                    RoundedRectangle(cornerRadius: 8, style: .continuous)
                        .strokeBorder(
                            on ? LiveDesign.accent.opacity(0.75) : LiveDesign.muted.opacity(0.75),
                            style: StrokeStyle(lineWidth: 1, dash: [3, 3])
                        )
                        .padding(-2)
                }
        } else {
            content
        }
    }
}

/// One eye badge. Tapping it shows or hides that element for the DISP mode being edited.
struct ChromeEditBadge: View {
    @Environment(AppModel.self) private var model
    let section: PocketDispChrome.Section
    let mode: PocketDispMode

    var body: some View {
        let on = model.chrome(for: mode).isVisible(section)
        Button {
            OperatorSettingsHaptics.selection(enabled: model.hapticsEnabled)
            model.toggleChrome(section, for: mode)
        } label: {
            (on ? OpcIcon.eye : OpcIcon.eyeOff)
                .frame(width: 11, height: 11)
                .foregroundStyle(on ? LiveDesign.background : LiveDesign.text)
                .frame(
                    width: PocketChromeEditLayout.badgeSize,
                    height: PocketChromeEditLayout.badgeSize
                )
                .background(on ? LiveDesign.accent : Color.black.opacity(0.9), in: Circle())
                .overlay(Circle().strokeBorder(LiveDesign.text.opacity(0.55), lineWidth: 1))
        }
        .buttonStyle(.zcTapTarget)
        .minTapTarget()
        .accessibilityLabel(section.title)
        .accessibilityValue(on ? "Shown" : "Hidden")
        .accessibilityHint("Shows or hides this on \(mode.title)")
        .accessibilityIdentifier("monitor.chromeEdit.\(section.rawValue)")
    }
}

struct ChromeEditBadgeLayer: View {
    @Environment(AppModel.self) private var model
    let mode: PocketDispMode
    let boxes: [PocketDispChrome.Section: CGRect]
    let viewport: CGSize

    var body: some View {
        let frames = PocketChromeEditLayout.badgeFrames(
            PocketDispChrome.Section.allCases.compactMap { section in
                boxes[section].map {
                    PocketChromeEditLayout.Box(section: section, frame: $0)
                }
            },
            viewport: viewport
        )
        ZStack(alignment: .topLeading) {
            ForEach(PocketDispChrome.Section.allCases) { section in
                if let frame = frames[section] {
                    ChromeEditBadge(section: section, mode: mode)
                        .environment(model)
                        .liveModuleFrame(frame)
                }
            }
        }
        .frame(width: viewport.width, height: viewport.height, alignment: .topLeading)
    }
}

/// Names the mode being edited and gets the operator back to Display settings.
struct ChromeEditBanner: View {
    @Environment(AppModel.self) private var model
    let mode: PocketDispMode

    var body: some View {
        HStack(spacing: 10) {
            OpcIcon.eye
                .frame(width: 11, height: 11)
                .foregroundStyle(LiveDesign.accent)
            VStack(alignment: .leading, spacing: 1) {
                Text("Editing \(mode.title)")
                    .font(LiveType.ui(size: 11.5, weight: .semibold))
                    .foregroundStyle(LiveDesign.text)
                Text("Tap an eye to show or hide it")
                    .font(LiveType.ui(size: 10, weight: .regular))
                    .foregroundStyle(LiveDesign.muted)
            }
            Button {
                model.endChromeEditing()
            } label: {
                Text("Done")
                    .font(LiveType.ui(size: 11.5, weight: .bold))
                    .foregroundStyle(LiveDesign.background)
                    .padding(.horizontal, 12)
                    .padding(.vertical, 6)
                    .background(LiveDesign.accent, in: Capsule())
            }
            .buttonStyle(.zcTapTarget)
            .accessibilityIdentifier("monitor.chromeEdit.done")
        }
        .padding(.leading, 12)
        .padding(.trailing, 6)
        .padding(.vertical, 6)
        .liquidGlass(in: Capsule())
        .accessibilityIdentifier("monitor.chromeEdit.banner")
    }
}
